spirv_glsl.cpp

```diff
@@ -718,6 +719,54 @@ void CompilerGLSL::emit_header()
 			statement("#extension ", ext, " : enable");
 			statement("#endif");
 		}
+		else if (ext == "GL_KHR_shader_subgroup_vote")
+		{
+			declaredKHR_shader_subgroup_vote = true;
+			if (options.vulkan_semantics)
+				statement("#extension GL_KHR_shader_subgroup_vote : require");
+			else
+			{
+				statement("#if defined(GL_ARB_shader_group_vote)");
+				statement("#extension GL_ARB_shader_group_vote : require");
+				statement("#elif defined(GL_NV_gpu_shader5)");
+				statement("#extension GL_NV_gpu_shader5 : require");
+				statement("#elif defined(GL_AMD_gcn_shader)&&defined(GL_AMD_gpu_shader_int64)");
+				statement("#extension GL_AMD_gcn_shader : require");
+				statement("#extension GL_AMD_gpu_shader_int64 : enable"); //needed?
+				statement("#else");
+				statement("#error No extensions available for subgroupAllEqual()");
+				statement("#endif");
+			}
+		}
+		else if (ext == "GL_KHR_shader_subgroup_basic")
+		{
+			declaredKHR_shader_subgroup_basic = true;
+			if (options.vulkan_semantics)
+				statement("#extension GL_KHR_shader_subgroup_basic : require");
+			else
+			{
+				statement("#if defined(GL_NV_shader_thread_group)");
+
+				statement("#define GL_NV_shader_thread_group");
+
+				statement("#elif defined(GL_ARB_shader_ballot)");
+
+				statement("#extension GL_ARB_shader_ballot : require");
+				statement("#extension GL_AMD_gcn_shader : enable");
+				statement("#ifdef GL_AMD_gcn_shader");
+				statement("#define gl_SubgroupSize uint(gl_SIMDGroupSizeAMD)");
+				statement("#else defined(GL_ARB_shader_ballot)");
+				statement("#define gl_SubgroupSize gl_SubGroupSizeARB");
+				statement("#endif");
+				statement("#define gl_SubgroupInvocationID gl_SubGroupInvocationARB");
+
+				statement("#endif");
+
+				statement("#define gl_SubgroupID (gl_LocalInvocationIndex/gl_SubgroupSize)");
+				statement("#define gl_NumSubgroups "
+				          "(gl_WorkGroupSize.x*gl_WorkGroupSize.y*gl_WorkGroupSize.z/gl_SubgroupSize)");
```

Comment:
This gets a bit sketchy. Workgroup size might not align with the subgroup size, and the rounding behavior is not exactly known to the application. E.g. a workgroup size of (5, 3, 7) is hard to know for sure what happens.

Reply:
Well for `GL_NV_shader_thread_group` we can define the following (@Crisspl the NV extension should always be our "primary"/preferred implementation)
```glsl
gl_SubgroupSize = gl_WarpSizeNV
gl_SubgroupInvocationID = gl_ThreadInWarpNV 
gl_SubgroupID = gl_WarpIDNV
gl_NumSubgroups = gl_WarpsPerSMNV
```

Reply:
on AMD/ARB the `gl_SubgroupSize` can be obtained from `gl_SIMDGroupSizeAMD` and `gl_SubGroupSizeARB` respectively.

Non-NV requires `ARB_shader_ballot` to obtain `gl_SubgroupInvocationID` which is `gl_SubGroupInvocationARB`

Reply:
You're right that `gl_NumSubgroup` and `gl_NumSubgroups` are impossible to obtain on non-NVidia systems, @HansKristian-Work